### PR TITLE
Support arbitrary length bit strings

### PR DESCRIPTION
**Pull request by mdeshon-google**
* Added second EncodeString prototype using vector<uint8_t> for >32 bit strings (byte-aligned)
* Added second MakeBloomFilter protoype using vector<uint8_t> for >32 bit strings
* Converted rappor::Encode to throw exceptions instead of using assert()
* Wrote unit tests, including one to ensure that old and new EncodeString work equivalently
* Still throw errors if attempting to use HmacSha256 with >32 bits

### CODE REVIEW

File: client/cpp/encoder.cc [hunk 1]
```diff
@@ -53,7 +54,7 @@ static void CheckValidProbability(float prob, const char* var_name) {
   if (prob < 0.0f || prob > 1.0f) {
     log("%s should be between 0.0 and 1.0 inclusive (got %.2f)", var_name,
         prob);
-    assert(false);
+    throw std::out_of_range("prob");
   }
```
Comment (ilyamironov) [inline client/cpp/encoder.cc]: Although this is an OSS project, we generally follow Google style guide, which frowns upon use of C++ exceptions.

File: client/cpp/encoder.cc [hunk 2]
```diff
@@ -83,9 +84,18 @@ uint32_t Encoder::AssignCohort(const Deps& deps, int num_cohorts) {
...
+  }
+
+  // Either we are using SHA256 to have exactly 32 bytes,
+  // or we're using HmacDrbg for any number of bytes.
+  if ((sha256.size() == kMaxBits)
+          || (deps.hmac_func_ == rappor::HmacDrbg)) {
+    // Hash size ok.
+  } else {
+    log("Bad hash size.");
+    throw std::out_of_range("sha256.size()");
```
Comment (ilyamironov) [inline client/cpp/encoder.cc]: Same comment about exceptions apply.

File: client/cpp/encoder.cc [hunk 3]
```diff
@@ -256,4 +331,89 @@ bool Encoder::EncodeString(const std::string& value, Bits* irr_out) const {
   return _EncodeStringInternal(value, &unused_bloom, &unused_prr, irr_out);
 }
 
+static uint8_t shifted(const Bits& bits, const int& index) {
+  // For an array of bytes, select the appopriate byte from a 4-byte
+  // integer value.
+  int result = 0;
+  int shift = 8 * (3 - (index % 4)); // Byte 0 shifts by 24 bits, 1 by 16, etc.
+  result = bits & (0xFF << shift);  // Select the appropriate byte
+  result = result >> shift;          // Shift to LSB
```
Comment (ilyamironov) [inline client/cpp/encoder.cc]: I'd would have written the last two lines as one: result = (bits >> shift) 0xFF;

File: client/cpp/encoder.cc [hunk 4]
```diff
@@ -256,4 +331,89 @@ bool Encoder::EncodeString(const std::string& value, Bits* irr_out) const {
   return _EncodeStringInternal(value, &unused_bloom, &unused_prr, irr_out);
 }
 
+static uint8_t shifted(const Bits& bits, const int& index) {
+  // For an array of bytes, select the appopriate byte from a 4-byte
+  // integer value.
```
Comment (ilyamironov) [inline client/cpp/encoder.cc]: It is important to document that bytes are enumerated in the big-endian order: index = 0 selects the most significant byte, index = 3 the least significant.

File: client/cpp/encoder.cc (hunk 1 above)
Comment (mdeshon-google) [inline client/cpp/encoder.cc]: OK, reverted back to using assert(false). I thought that was only an internal thing, but now I see the external style guide also says not to use exceptions. Oh, well.

File: client/cpp/encoder.cc (hunk 2 above)
Comment (mdeshon-google) [inline client/cpp/encoder.cc]: Done everywhere.

File: client/cpp/encoder.cc (hunk 3 above)
Comment (mdeshon-google) [inline client/cpp/encoder.cc]: Compactified.

File: client/cpp/encoder.cc (hunk 4 above)
Comment (mdeshon-google) [inline client/cpp/encoder.cc]: Done.

File: client/cpp/encoder.cc (hunk 1 above)
Comment (mdeshon-google) [inline client/cpp/encoder.cc]: Also added a bunch of new negative tests.

**Pull request merged**